vacasaconnect.py

```diff
@@ -0,0 +1,243 @@
+                 language: str = 'en-US',
+                 currency: str = 'USD'
+                 ):
+        """Initialize an instance of the VacasaConnect class
+
+        Args:
+            endpoint: The URL of the Vacasa Connect API.
+            api_key: Your Vacasa Connect API key.
+            api_secret: Your Vacasa Connect API secret.
+            timezone: UTC or a long-form version of a timezone from the tz
+                database. Example: 'America/New_York'. See
+                https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
+                for more info.
+            language: The language you will accept. If no localized content
+                can be found, an error response is returned. Allowed format is
+                {{ISO-639-1 Code}}-{{ISO 3166-1 Alpha-2 Code}}.
+            currency: An ISO-4217 currency code. Send to request monetary
+                values in this currency.
+        """
+        self.endpoint = endpoint.rstrip('/')
+        self.api_key = api_key
+        self.api_secret = api_secret
+        self.timezone = timezone
+        self.language = language
+        self.currency = currency
+        self._populate_tokens()
+
+    def _populate_tokens(self):
+        """Decide if current tokens need refreshed and populate them"""
+        # generate a token if we don't have one yet
+        if self._access_token is None:
+            tokens = self._get_new_tokens()
+            self._access_token = tokens.get('access_token')
+            self._refresh_token = tokens.get('refresh_token')
+        else:
+            now = pendulum.now()
+            expiration = pendulum.parse(self._refresh_token.get('expires_at'))
+
+            # refresh the token if it has expired
+            if now > expiration:
+                tokens = self._refresh_tokens()
+                self._access_token = tokens.get('access_token')
+                self._refresh_token = tokens.get('refresh_token')
+
+    def _get_new_tokens(self) -> dict:
+        """Generate new access and refresh tokens"""
+        timestamp = int(pendulum.now().timestamp())
+        payload = json.dumps({
+            'data': {
+                'api_key': self.api_key,
+                'timestamp': timestamp,
+                'signature': self._generate_signature(timestamp)
+            }
+        })
+        headers = {'content-type': 'application/json'}
+
+        r = self._post(f"{self.endpoint}/auth", payload, headers)
+        tokens = r.json().get('data').get('attributes')
+        self._validate_tokens(tokens)
+
+        return tokens
+
+    def _refresh_tokens(self) -> dict:
+        """Refresh existing tokens. Necessary when they expire."""
+        url = f"{self.endpoint}/auth/refresh"
+        payload = {
+            'data': {
+                'refresh_token': self._refresh_token.get('token')
+            }
+        }
+        r = self._post(url, payload)
+        tokens = r.json().get('data').get('attributes')
```

Comment:
This might flow to validation a little more smoothly like this, maybe?
```python
tokens = r.json().get('data', {}).get('attributes', {})
```

I think there's a weird edge case here where you'll get unhandled exceptions if the json comes back in an unexpected form.

Reply:
Yep. I will change that. Thanks!